internal/rust/semantic/src/types.rs

```diff
@@ -0,0 +1,705 @@
+use std::{cmp, collections::HashMap, fmt};
+
+// PolyType represents a generic parametrized type.
+//
+// TODO:
+//     Do not derive PartialEq implementation.
+//     Instead provide a custom implementation
+//     that instantiates both polytypes with the
+//     same type variables.
+//
+//     Note this requires a substitution, so remove
+//     this derivation once substitutions are defined.
+//
+#[derive(Debug, Clone, PartialEq)]
+pub struct PolyType {
+    pub free: Vec<Tvar>,
+    pub bnds: Option<HashMap<Tvar, Kind>>,
+    pub expr: MonoType,
+}
+
+impl fmt::Display for PolyType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self.bnds {
+            Some(bnds) => {
+                let mut bounds = Vec::new();
+                for tv in &self.free {
+                    if let Some(kind) = bnds.get(tv) {
+                        bounds.push(BoundTvar {
+                            tv: *tv,
+                            kind: *kind,
+                        })
+                    }
+                }
+                write!(
+                    f,
+                    "forall [{}] where {} {}",
+                    DisplayList {
+                        values: &self.free,
+                        delim: ", "
+                    },
+                    DisplayList {
+                        values: &bounds,
+                        delim: ", "
+                    },
+                    self.expr,
+                )
+            }
+            None => write!(
+                f,
+                "forall [{}] {}",
+                DisplayList {
+                    values: &self.free,
+                    delim: ", "
+                },
+                self.expr
+            ),
+        }
+    }
+}
+
+// Kind represents a class or family of types
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Kind {
+    Addable,
+    Subtractable,
+    Divisible,
+    Comparable,
+    Nullable,
+}
+
+impl fmt::Display for Kind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Kind::Addable => f.write_str("Addable"),
+            Kind::Subtractable => f.write_str("Subtractable"),
+            Kind::Divisible => f.write_str("Divisible"),
+            Kind::Comparable => f.write_str("Comparable"),
+            Kind::Nullable => f.write_str("Nullable"),
+        }
+    }
+}
+
+// MonoType represents a specific named type
+#[derive(Debug, Clone, PartialEq)]
+pub enum MonoType {
+    Bool,
+    Int,
+    Uint,
+    Float,
+    Str,
```

Comment:
nit, to be consistent this should be String.

Reply:
Oops I thought the name `String` was reserved, but it's not.